### PR TITLE
Various battle fixes

### PR DESCRIPTION
**Pull request by fmatthew5876**
see commits

Fix: #2102
Fix: #2123

### CODE REVIEW

**Comment by fmatthew5876** [comment]
Tested #2102 with this patch. In master an all enemy skill hits all the hidden birds. In this PR it only hits the visible one.

File: src/game_enemy.cpp [hunk 1]
```diff
@@ -169,6 +169,7 @@ int Game_Enemy::GetHue() const {
 
 void Game_Enemy::SetHidden(bool _hidden) {
 	hidden = _hidden;
+	ResetBattleStateOnDeath();
```
Comment (Ghabry) [inline src/game_enemy.cpp:172]: I don't think SetHidden should change anything. They are just hidden and keep everything.
Because this is unconditionally enemies have 0 HP when the battle starts.

Also you forgot the edge case (bug?) that enemies removed via "Force Flee" in 2k3 can reappear with "Enemy encounter", another reason against "SetHp(0)" ;)

**Comment by fmatthew5876** [comment]
Test Cases
========

1.
--------
Start enemy hidden
Turn 0 event inflict poison
Turn 2 event unhide

Result:
Turn 1: enemy appears
Turn 2: enemy takes poison damage

2.
---------------

Set enemy action Always Defend 50%, *3+0 Self-Destruct 100%
Turn *3+1 event: Show enemy1

Turn 1: Cast fire on enemy - Normal damage
Turn 2: Use skill which inflicts poison and reduces fire element
Turn 3: Enemy self destructs
Turn 4: Enemy re-appears
Turn 5: Cast fire on enemy - High damage
Turn 5: Enemy takes poison damage


Conclusions
------------

Hide doesn't appear to change anything about enemies. They just stop participating in the battle.

* Showing an enemy does not change their states

**Comment by fmatthew5876** [comment]
Test Cases Add/Rem party
-------------------------------

Battle event 1 triggered by SW1 on
```
Text "Remove"
Remove Brian
Sw1 Off
```

Battle event 2 triggered by SW2 on
```
Text "Add"
Add Brian
Sw2 Off
```

1.
------

Turn 1: Brian Defends
Turn 1: Alex uses skill which turns on Sw1
Turn 1: "Remove" - Brian removed
Turn 1: Carol uses skill which turns on Sw2
Turn 1: "Add" - Brian Added
Turn 1: Enemy uses physical skill which hits all
Turn 1: Brian takes half damage :grey_exclamation: 

2. 
--------
Turn 1: Brian Defends
Turn 1: Alex uses skill which turns on Sw1
Turn 1: "Remove" - Brian removed
Turn 1: Carol Defends
Turn 1: Enemy uses physical skill which hits all

Turn 2: Alex Defends
Turn 2: Carol uses skill which turns on Sw2
Turn 2: "Add" - Brian Added
Turn 2: Enemy uses physical skill which hits all
Turn 2: Brian takes half damage :exclamation: 

3.
-----

Turn 1: Brian uses a skill to increase fire defense on himself
Turn 1: Alex uses skill to remove brian

Turn 2: Carol uses skill to add brian
Turn 2: Enemy uses fire magic on all
Turn 2: Brian takes half damage

4. 
---
Turn 1: Alex uses a skill to increase fire defense of Brian
Turn 1: Carol uses skill to remove brian
Win the battle

Battle again
Turn 1: Alex uses a skill to add brian
Turn 1: Enemy uses fire magic on all
Turn 1: Brian takes normal damage

Conclusion
---------------

* Add/Remove from party doesn't change any battle states, including defend!
* Between battles, Battle states are cleared for all actors, even those outside of the party

**Comment by fmatthew5876** [comment]
This has been cleaned up and tested.

I was able to come up with a very nice way to handle defend and charged states. 

Now every time we construct a new battle algorithm, we remove both the states in `AlgorithmBase::AlgorithmBase()`. For charged, we carry the charge flag over into the algo itself. For defend, we set the defend flag in Defend's constructor, which gets called after the base one.

This works for 2k and works very nicely for 2k3. In 2k3 it means you won't come out of defend state until you select an action, which is what RPG_RT does.

**Comment by fmatthew5876** [comment]
Added the party fixes and tested all the edge cases.

**Comment by Ghabry** [comment]
Thanks. Will review this soon. (I consider this important enough to get it merged before the next Android apk push, because then we have everything in).

File: src/game_battlealgorithm.cpp [hunk 2]
```diff
@@ -114,6 +114,9 @@ Game_BattleAlgorithm::AlgorithmBase::AlgorithmBase(Type ty, Game_Battler* source
 	source_restriction(RPG::State::Restriction(source->GetSignificantRestriction()))
 {
 	Reset();
+	if (source) {
```
Comment (Ghabry) [inline src/game_battlealgorithm.cpp:117]: > Turn off defend status in AlgorithmBase constructor 

The source check is unnecessary. source is never null (there is a source-> access a few lines above btw)

Comment (fmatthew5876) [reply src/game_battlealgorithm.cpp]: Removed it

**Comment by Ghabry** [comment]
There is a suspicious call to ``actor->ResetBattle()`` in ``Interpreter::CommandSimulatedAttack``.
Intention is likely to get rid of battle modifiers because of ``getDef`` and ``getSpi``

**Comment by fmatthew5876** [comment]
> There is a suspicious call to `actor->ResetBattle()` in `Interpreter::CommandSimulatedAttack`.
> Intention is likely to get rid of battle modifiers because of `getDef` and `getSpi` but this event command can be called during battle :/

I saw that and wasn't sure. Since we're properly cleaning up before and after battle, it's probably safe to remove now.
